File: packages/js/src/types/built-in.ts
/**
 * Built-in artifact types for the Codex SDK
 *
 * These types define common artifact categories with default TTLs,
 * archival policies, and pattern matching rules.
 */

/**
 * Artifact type definition
 */
export interface ArtifactType {
  /** Unique type name */
  name: string
  /** Human-readable description */
  description: string
  /** Glob patterns to match files of this type */
  patterns: string[]
  /** Default TTL in seconds for cached content */
  defaultTtl: number
  /** Days after which to archive (null = never) */
  archiveAfterDays: number | null
  /** Where to archive (null = don't archive) */
  archiveStorage: 'local' | 'cloud' | 'drive' | null
  /** Patterns for sync operations (optional override) */
  syncPatterns?: string[]
  /** Patterns to exclude from sync */
  excludePatterns?: string[]
  /** Default permissions (optional) */
  permissions?: {
    include: string[]
    exclude: string[]
  }
}

/**
 * TTL constants in seconds
 */
export const TTL = {
  ONE_HOUR: 3600,
  ONE_DAY: 86400,
  ONE_WEEK: 604800,
  TWO_WEEKS: 1209600,
  ONE_MONTH: 2592000,
  ONE_YEAR: 31536000,
} as const

/**
 * Built-in artifact types
 *
 * These cover common use cases and serve as defaults.
 * Users can override or extend these via configuration.
 */
export const BUILT_IN_TYPES: Record<string, ArtifactType> = {
  docs: {
    name: 'docs',
    description: 'Documentation files',
    patterns: ['docs/**', 'README.md', 'CLAUDE.md', '*.md'],
    defaultTtl: TTL.ONE_WEEK,
    archiveAfterDays: 365,
    archiveStorage: 'cloud',
  },

  specs: {
    name: 'specs',
    description: 'Technical specifications',
    patterns: ['specs/**', 'SPEC-*.md', '**/specs/**'],
    defaultTtl: TTL.TWO_WEEKS,
    archiveAfterDays: null, // Never auto-archive specs
    archiveStorage: null,
  },

  logs: {
    name: 'logs',
    description: 'Session and workflow logs',
    patterns: ['.fractary/**/logs/**', 'logs/**', '*.log'],
    defaultTtl: TTL.ONE_DAY,
    archiveAfterDays: 30,
    archiveStorage: 'cloud',
  },

  standards: {
    name: 'standards',
    description: 'Organization standards and guides',
    patterns: ['standards/**', 'guides/**', '.fractary/standards/**'],
    defaultTtl: TTL.ONE_MONTH,
    archiveAfterDays: null, // Never auto-archive standards
    archiveStorage: null,
  },

  templates: {
    name: 'templates',
    description: 'Reusable templates',
    patterns: ['templates/**', '.templates/**', '.fractary/templates/**'],
    defaultTtl: TTL.TWO_WEEKS,
    archiveAfterDays: 180,
    archiveStorage: 'cloud',
  },

  state: {
    name: 'state',
    description: 'Workflow state files',
    patterns: ['.fractary/**/state.json', '.fractary/**/state/**', '**/state.json'],
    defaultTtl: TTL.ONE_HOUR,
    archiveAfterDays: 7,
    archiveStorage: 'local',
  },

  config: {
    name: 'config',
    description: 'Configuration files',
    patterns: [
      '.fractary/**/*.json',
      '*.config.js',
      '*.config.ts',
      'tsconfig.json',
      'package.json',
    ],
    defaultTtl: TTL.ONE_DAY,
    archiveAfterDays: null,
    archiveStorage: null,
  },

  systems: {
    name: 'systems',
    description: 'System definitions',
    patterns: ['.fractary/systems/**', 'systems/**'],
    defaultTtl: TTL.TWO_WEEKS,
    archiveAfterDays: null,
    archiveStorage: null,
  },
}

/**
 * Default type to use when no pattern matches
 */
export const DEFAULT_TYPE: ArtifactType = {
  name: 'default',
  description: 'Default type for unmatched files',
  patterns: ['**/*'],
  defaultTtl: TTL.ONE_WEEK,
  archiveAfterDays: 90,
  archiveStorage: 'cloud',
}

/**
 * Get a built-in type by name
 *
 * @param name - Type name
 * @returns The type or undefined
 */
export function getBuiltInType(name: string): ArtifactType | undefined {
  return BUILT_IN_TYPES[name]
}

/**
 * Get all built-in type names
 *
 * @returns Array of type names
 */
export function getBuiltInTypeNames(): string[] {
  return Object.keys(BUILT_IN_TYPES)
}

/**
 * Check if a name is a built-in type
 *
 * @param name - Type name to check
 * @returns true if it's a built-in type
 */
export function isBuiltInType(name: string): boolean {
  return name in BUILT_IN_TYPES
}

File: sdk/py/fractary_codex/types/builtin.py
"""
Built-in artifact types and TTL constants.

This module defines the default artifact types recognized by the Codex SDK,
along with their default TTL (time-to-live) values for caching.
"""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Optional


class TTL:
    """Time-to-live constants in seconds.

    These constants provide human-readable TTL values for common durations.

    Examples:
        >>> from fractary_codex.types import TTL
        >>> cache_ttl = TTL.HOUR * 2  # 2 hours
        >>> long_ttl = TTL.WEEK
    """

    MINUTE: int = 60
    HOUR: int = 3600
    DAY: int = 86400
    WEEK: int = 604800
    MONTH: int = 2592000  # 30 days
    YEAR: int = 31536000  # 365 days

    # Common aliases
    FIVE_MINUTES: int = 300
    FIFTEEN_MINUTES: int = 900
    THIRTY_MINUTES: int = 1800
    SIX_HOURS: int = 21600
    TWELVE_HOURS: int = 43200
    TWO_DAYS: int = 172800
    THREE_DAYS: int = 259200


@dataclass(frozen=True)
class ArtifactType:
    """Definition of an artifact type.

    Attributes:
        name: Unique identifier for the type (e.g., 'docs', 'config')
        patterns: Glob patterns that match this type
        ttl: Default TTL in seconds for caching
        description: Human-readable description
        priority: Priority for pattern matching (higher = checked first)
    """

    name: str
    patterns: Sequence[str]
    ttl: int
    description: str = ""
    priority: int = 0


# Built-in artifact types with default TTLs
# These match the TypeScript SDK's built-in types
BUILT_IN_TYPES: dict[str, ArtifactType] = {
    "docs": ArtifactType(
        name="docs",
        patterns=["docs/**/*.md", "docs/**/*.mdx", "*.md"],
        ttl=TTL.DAY,
        description="Documentation files",
        priority=10,
    ),
    "config": ArtifactType(
        name="config",
        patterns=[
            "*.yaml",
            "*.yml",
            "*.json",
            "*.toml",
            ".fractary/**",
            "config/**",
        ],
        ttl=TTL.HOUR,
        description="Configuration files",
        priority=20,
    ),
    "schema": ArtifactType(
        name="schema",
        patterns=[
            "schemas/**/*.json",
            "schemas/**/*.yaml",
            "**/*.schema.json",
            "**/*.schema.yaml",
        ],
        ttl=TTL.DAY,
        description="Schema definition files",
        priority=30,
    ),
    "templates": ArtifactType(
        name="templates",
        patterns=["templates/**/*", "**/*.template.*", "**/*.tmpl"],
        ttl=TTL.WEEK,
        description="Template files",
        priority=15,
    ),
    "specs": ArtifactType(
        name="specs",
        patterns=["specs/**/*.md", "specifications/**/*.md"],
        ttl=TTL.DAY,
        description="Specification documents",
        priority=25,
    ),
    "workflows": ArtifactType(
        name="workflows",
        patterns=[".github/workflows/**/*.yaml", ".github/workflows/**/*.yml"],
        ttl=TTL.SIX_HOURS,
        description="CI/CD workflow files",
        priority=40,
    ),
    "scripts": ArtifactType(
        name="scripts",
        patterns=["scripts/**/*.sh", "scripts/**/*.py", "bin/**/*"],
        ttl=TTL.TWELVE_HOURS,
        description="Script files",
        priority=35,
    ),
    "prompts": ArtifactType(
        name="prompts",
        patterns=["prompts/**/*.md", "prompts/**/*.txt", "**/*.prompt.md"],
        ttl=TTL.HOUR,
        description="AI prompt templates",
        priority=50,
    ),
    "agents": ArtifactType(
        name="agents",
        patterns=["agents/**/*", "**/*.agent.yaml", "**/*.agent.md"],
        ttl=TTL.HOUR,
        description="AI agent definitions",
        priority=45,
    ),
    "skills": ArtifactType(
        name="skills",
        patterns=["skills/**/*", "**/*.skill.yaml", "**/*.skill.md"],
        ttl=TTL.HOUR,
        description="AI skill definitions",
        priority=45,
    ),
}


# Default TTL when no type matches
DEFAULT_TTL: int = TTL.DAY


def get_built_in_type(name: str) -> Optional[ArtifactType]:
    """Get a built-in type by name.

    Args:
        name: Type name (e.g., 'docs', 'config')

    Returns:
        ArtifactType if found, None otherwise

    Examples:
        >>> t = get_built_in_type("docs")
        >>> t.ttl
        86400
    """
    return BUILT_IN_TYPES.get(name)


def get_all_built_in_types() -> list[ArtifactType]:
    """Get all built-in types sorted by priority (highest first).

    Returns:
        List of ArtifactType sorted by priority descending
    """
    return sorted(BUILT_IN_TYPES.values(), key=lambda t: -t.priority)
